packages/cursorless-vscode/src/CommandHistory.ts

```diff
@@ -0,0 +1,147 @@
+import {
+  ActionDescriptor,
+  CommandComplete,
+  CommandHistoryItem,
+  Disposable,
+  FileSystem,
+  ReadOnlyHatMap,
+} from "@cursorless/common";
+import type {
+  CommandRunner,
+  CommandRunnerDecorator,
+} from "@cursorless/cursorless-engine";
+import * as fs from "node:fs/promises";
+import * as path from "node:path";
+import * as vscode from "vscode";
+
+const filePrefix = "cursorlessCommandHistory";
+const settingSection = "cursorless";
+const settingName = "commandHistory";
+const settingFullName = `${settingSection}.${settingName}`;
+
+export class CommandHistory implements CommandRunnerDecorator {
+  private readonly dirPath: string;
+  private readonly cursorlessVersion: string;
+  private disposable: Disposable;
+  private isActive: boolean = false;
+
+  constructor(
+    extensionContext: vscode.ExtensionContext,
+    fileSystem: FileSystem,
+  ) {
+    this.cursorlessVersion = extensionContext.extension.packageJSON.version;
+    this.dirPath = fileSystem.cursorlessCommandHistoryDirPath;
+
+    // Read initial setting value. The watcher below will take care of changes.
+    this.evaluateSetting();
+
+    this.disposable = vscode.workspace.onDidChangeConfiguration((event) => {
+      if (event.affectsConfiguration(settingFullName)) {
+        this.evaluateSetting();
+      }
+    });
+  }
+
+  wrapCommandRunner(
+    readableHatMap: ReadOnlyHatMap,
+    runner: CommandRunner,
+  ): CommandRunner {
+    if (!this.isActive) {
+      return runner;
+    }
+
+    return {
+      run: async (commandComplete: CommandComplete) => {
+        await this.appendToLog(commandComplete);
+
+        return await runner.run(commandComplete);
+      },
+    };
+  }
+
+  private async appendToLog(command: CommandComplete): Promise<void> {
+    const date = new Date();
+    const fileName = `${filePrefix}_${getMonthDate(date)}.jsonl`;
+    const file = path.join(this.dirPath, fileName);
+
+    const historyItem: CommandHistoryItem = {
+      date: getDayDate(date),
+      cursorlessVersion: this.cursorlessVersion,
+      command: sanitizeCommand(command),
+    };
+    const data = JSON.stringify(historyItem) + "\n";
+
+    await fs.mkdir(this.dirPath, { recursive: true });
+    await fs.appendFile(file, data, "utf8");
+  }
+
+  private evaluateSetting() {
+    this.isActive = vscode.workspace
+      .getConfiguration(settingSection)
+      .get<boolean>(settingName, false);
+  }
+
+  dispose() {
+    this.disposable.dispose();
+  }
+}
+
+// Remove spoken form and sanitize action
+function sanitizeCommand(command: CommandComplete): CommandComplete {
+  const { spokenForm, action, ...rest } = command;
+  return {
+    ...rest,
+    action: sanitizeAction(action),
+  };
+}
+
+function sanitizeAction(action: ActionDescriptor): ActionDescriptor {
+  switch (action.name) {
```

Comment:
I know it's a bit annoying, but I would be tempted to force us to enumerate all actions here. Otherwise we may inadvertently start logging sensitive information in the future

Reply:
updated

Reply:
Yeah I think generally we should try to avoid them if we can, but in this case I think it's justified